packages/atclient/src/metadata.c

```diff
@@ -677,263 +681,377 @@ exit: {
+  long len = 0;
+  if (atclient_atkey_metadata_is_ttl_initialized(metadata)) {
+    len += atclient_atkey_metadata_ttl_strlen(metadata);
+  }
 
-  atclient_atstr protocolstr;
-  atclient_atstr_init(&protocolstr, 4096);
+  if (atclient_atkey_metadata_is_ttb_initialized(metadata)) {
+    len += atclient_atkey_metadata_ttb_strlen(metadata);
+  }
+
+  if (atclient_atkey_metadata_is_ttr_initialized(metadata)) {
+    len += atclient_atkey_metadata_ttr_strlen(metadata);
+  }
+
+  if (atclient_atkey_metadata_is_ccd_initialized(metadata) && metadata->ccd == 1) {
+    len += atclient_atkey_metadata_ccd_strlen(metadata);
+  }
+
+  if (atclient_atkey_metadata_is_isbinary_initialized(metadata) && metadata->isbinary) {
+    len += atclient_atkey_metadata_isbinary_strlen(metadata);
+  }
+
+  if (atclient_atkey_metadata_is_isencrypted_initialized(metadata) && metadata->isencrypted) {
+    len += atclient_atkey_metadata_isencrypted_strlen(metadata);
+  }
+
+  if (atclient_atkey_metadata_is_datasignature_initialized(metadata)) {
+    len += atclient_atkey_metadata_datasignature_strlen(metadata);
+  }
+
+  if (atclient_atkey_metadata_is_sharedkeystatus_initialized(metadata)) {
+    len += atclient_atkey_metadata_sharedkeystatus_strlen(metadata);
+  }
+
+  if (atclient_atkey_metadata_is_sharedkeyenc_initialized(metadata)) {
+    len += atclient_atkey_metadata_sharedkeyenc_strlen(metadata);
+  }
+
+  if (atclient_atkey_metadata_is_pubkeyhash_initialized(metadata)) {
+    len += atclient_atkey_metadata_pubkeyhash_strlen(metadata);
+  }
+
+  if (atclient_atkey_metadata_is_pubkeyalgo_initialized(metadata)) {
+    len += atclient_atkey_metadata_pubkeyalgo_strlen(metadata);
+  }
+
+  if (atclient_atkey_metadata_is_encoding_initialized(metadata)) {
+    len += atclient_atkey_metadata_encoding_strlen(metadata);
+  }
+
+  if (atclient_atkey_metadata_is_enckeyname_initialized(metadata)) {
+    len += atclient_atkey_metadata_enckeyname_strlen(metadata);
+  }
+
+  if (atclient_atkey_metadata_is_encalgo_initialized(metadata)) {
+    len += atclient_atkey_metadata_encalgo_strlen(metadata);
+  }
+
+  if (atclient_atkey_metadata_is_ivnonce_initialized(metadata)) {
+    len += atclient_atkey_metadata_ivnonce_strlen(metadata);
+  }
+
+  if (atclient_atkey_metadata_is_skeenckeyname_initialized(metadata)) {
+    len += atclient_atkey_metadata_skeenckeyname_strlen(metadata);
+  }
+
+  if (atclient_atkey_metadata_is_skeencalgo_initialized(metadata)) {
+    len += atclient_atkey_metadata_skeencalgo_strlen(metadata);
+  }
+
+  return len;
+}
+
+size_t atclient_atkey_metadata_ttl_strlen(const atclient_atkey_metadata *metadata) {
+  return 5 // :ttl:
+         + long_strlen(metadata->ttl);
+}
+
+size_t atclient_atkey_metadata_ttb_strlen(const atclient_atkey_metadata *metadata) {
+  return 5 // :ttb:
+         + long_strlen(metadata->ttb);
+}
+
+size_t atclient_atkey_metadata_ttr_strlen(const atclient_atkey_metadata *metadata) {
+  return 5 // :ttr:
+         + long_strlen(metadata->ttr);
+}
```

Comment:
These functions are messy, but a necessary evil to reduce string buffering sizes.